### DISCUSSION TITLE
[Idea] Crop face/entropy

### DISCUSSION THREAD

**Discussion by masterix21** [spatie/image#114] [Ideas]
Hi guys, this is another great Spatie's package. Thank you for all your work.

Using it, I had the idea to integrate two new manipulations:
- [ ] Manipulations::CROP_ENTROPY
- [ ]  Manipulations::CROP_FACE

**Manipulations::CROP_FACE**
Detect the first face and crop the image centered on it.

**Manipulations::CROP_ENTROPY**
Usually, the image part with the major pixel disorder is the most relevant. It crops, centering the disorder.

If you like it, I will work on PR.

Thanks

**Comment by freekmurze** [comment]
This is an excellent idea. Feel free to investigate this further. 

Image you have a picture of two people sitting side by side. On the right we have Obama, on the left we have Trump. I'm thinking there should be some way of choosing on which face to focus on, otherwise some might think there is fraud going on in the election.

It would also be awesome if there would be a method that return the coordinates which were used for the crop. Those coordinates could be stored in a package like medialibrary and be used when generation conversions. But lets start with getting it right in this Image package first.

Thanks for your time on this, I'm excited to see what you come up with!

**Comment by masterix21** [comment]
> Image you have a picture of two people sitting side by side. On the right we have Obama, on the left we have Trump. I'm thinking there should be some way of choosing on which face to focus on, otherwise some might think there is fraud going on in the election.

Hahaha, yes, it's true. It could be an improvement after the first release.

> It would also be awesome if there would be a method that return the coordinates which were used for the crop. Those coordinates could be stored in a package like medialibrary and be used when generation conversions. But lets start with getting it right in this Image package first.

Good idea 💅

**Comment by masterix21** [comment]
@freekmurze, I lost my days thinking about these new features. Finally, I came to the conclusions I'm sharing with you.

**intervention/image**
It doesn't support features like "edge enhancement", "black threshold" and others: are available only with Imagemagick, no GD. I'm currently working on a PR to support all these features, but I don't know when and if they will accept it.

**Package manipulations**
This package has a lot of manipulations inside the class `Manipulations.php`. My idea is to create a manipulation interface implemented by any "real" manipulation (entropy, face, crop top right, etc), detaching all manipulations from a single class: creating a code easier to maintain. It means that the package could break to work and should be necessary to flag a new version, like 2.x.

I'm happy to share with you all these thoughts, to understand your philosophy about it.

**Comment by freekmurze** [comment]
Appreciate the update!

> creating a code easier to maintain

I'm all for that! 👍
